test/1_stdlib/StringAPI.swift

```diff
@@ -0,0 +1,439 @@
+  // U+0341 has a canonical decomposition mapping of U+0301.
+  ComparisonTest(.EQ, "\u{0301}", "\u{0341}"),
+  ComparisonTest(.LT, "\u{0301}", "\u{0954}",
+    xfail: .Custom({ objCPresent == false },
+      reason: "Compares as equal with ICU")),
+  ComparisonTest(.LT, "\u{0341}", "\u{0954}",
+    xfail: .Custom({ objCPresent == false },
+      reason: "Compares as equal with ICU")),
+]
+
+func checkStringComparison(
+  expected: ExpectedComparisonResult,
+  _ lhs: String, _ rhs: String, _ stackTrace: SourceLocStack
+) {
+  // String / String
+  expectEqual(expected.isEQ(), lhs == rhs, stackTrace: stackTrace)
+  expectEqual(expected.isNE(), lhs != rhs, stackTrace: stackTrace)
+  checkHashable(
+    expected.isEQ(), lhs, rhs, stackTrace: stackTrace.withCurrentLoc())
+
+  expectEqual(expected.isLT(), lhs < rhs, stackTrace: stackTrace)
+  expectEqual(expected.isLE(), lhs <= rhs, stackTrace: stackTrace)
+  expectEqual(expected.isGE(), lhs >= rhs, stackTrace: stackTrace)
+  expectEqual(expected.isGT(), lhs > rhs, stackTrace: stackTrace)
+  checkComparable(expected, lhs, rhs, stackTrace: stackTrace.withCurrentLoc())
+
+#if _runtime(_ObjC)
+  // NSString / NSString
+  let lhsNSString = lhs as NSString
+  let rhsNSString = rhs as NSString
+  let expectedEqualUnicodeScalars =
+    Array(lhs.unicodeScalars) == Array(rhs.unicodeScalars)
+  // FIXME: Swift String and NSString comparison may not be equal.
+  expectEqual(
+    expectedEqualUnicodeScalars, lhsNSString == rhsNSString,
+    stackTrace: stackTrace)
+  expectEqual(
+    !expectedEqualUnicodeScalars, lhsNSString != rhsNSString,
+    stackTrace: stackTrace)
+  checkHashable(
+    expectedEqualUnicodeScalars, lhsNSString, rhsNSString,
+    stackTrace: stackTrace.withCurrentLoc())
+#endif
+}
+
+for test in comparisonTests {
+  StringTests.test("String.{Equatable,Hashable,Comparable}: line \(test.loc.line)")
+  .xfail(test.xfail)
+  .code {
+    checkStringComparison(
+      test.expectedUnicodeCollation, test.lhs, test.rhs,
+      test.loc.withCurrentLoc())
+    checkStringComparison(
+      test.expectedUnicodeCollation.flip(), test.rhs, test.lhs,
+      test.loc.withCurrentLoc())
+  }
+}
+
+func checkCharacterComparison(
+  expected: ExpectedComparisonResult,
+  _ lhs: Character, _ rhs: Character, _ stackTrace: SourceLocStack
+) {
+  // Character / Character
+  expectEqual(expected.isEQ(), lhs == rhs, stackTrace: stackTrace)
+  expectEqual(expected.isNE(), lhs != rhs, stackTrace: stackTrace)
+  checkHashable(
+    expected.isEQ(), lhs, rhs, stackTrace: stackTrace.withCurrentLoc())
+
+  expectEqual(expected.isLT(), lhs < rhs, stackTrace: stackTrace)
+  expectEqual(expected.isLE(), lhs <= rhs, stackTrace: stackTrace)
+  expectEqual(expected.isGE(), lhs >= rhs, stackTrace: stackTrace)
+  expectEqual(expected.isGT(), lhs > rhs, stackTrace: stackTrace)
+  checkComparable(expected, lhs, rhs, stackTrace: stackTrace.withCurrentLoc())
+}
+
+for test in comparisonTests {
```

Comment:
I thought this would be better than duplicating the tests just for xfail cases.
Thanks for reviewing.